### ISSUE TITLE
Efficient Record Access

### ISSUE THREAD

**Issue by b-scholz** [souffle-lang/souffle#47]
Example: 

```
A(x) :- B(x), C([x, _]).
```

A full scan is performed for table `C`, i.e., all tuples of C are scrutinised to check whether the first element of the record is `x`.  To overcome this performance bottleneck, construct a helper relation that relates a record value with the record id. E.g., we can rewrite the example to

```
  C_helper(x) :- C([x,_]), B(x).
  A(x) :- B(x), C_helper(x).
```

**Comment by HerbertJordan** [comment]
Flattening records would be another, probably more efficient approach.

The declaration of 
```
C( x : [ a : number ,b : number ] ) 
```
should be translated into a relation
```
C( xa : number, xb : number )
```
and the indexing will be fixed without creating atom-specific helper relations. 

For recursive relations we could scan the full set of rules and look for the deepest nesting level of any clause in an atom. This should then be the unfolding degree.

For instance, in the code
```
.type List = [ a : number , r : List ]
.decl B ( a : number )
.decl C ( l : list ) 
.decl D ( l : list ) 

C ( r ) :- B(X) , C( [ X , r ] ).
D ( r ) :- C ( r ).
```
C would be rewritten to
```
.decl C ( a : number , r : List )
C ( ra , rr ) : - B(X), C ( X , [ ra, rr ] ).
D ( [ ra, rr ] ) :- D ( ra, rr ).
```
If there are even deeper matching rules, we would expand it even further.

**Comment by HerbertJordan** [comment]
A completely other solution would be to support hierarchical indexing. So, instead of having an index on attributes [2,1,3] of a relation we would also support an index on [2.1,1.2,3] where the first component of the second attribute is followed by the second component of the first attribute and the third attribute. This would avoid the rule expansion and put the workload in the indices.

**Comment by steve-s** [comment]
Flattening records: wouldn't it mean that the relation would have larger cardinality and so, for example, the decision between trie vs b-tree would be different? Moreover, can cardinality of relation have an effect to search performance?

**Comment by HerbertJordan** [comment]
There are many things to consider, but the fact that we can have an index on it + that data is accessed in a cache friendly way, not going through an indirection table of the records, should have massive performance benefits.

**Comment by steve-s** [comment]
Absolutely, I am trying to point out that records themselves may have other benefits so flattening is not a clear win seems to me.

**Comment by yanniss** [comment]
Has there been any progress on efficient record value access? It is a problem for us (Doop port, new research) on several occasions:

-The issue is not as simple as in the example Bernhard gave. Even insertion is hugely inefficient. E.g., I have a rule, perfectly optimized:
  ComputedAccessPath([?base2,[nil,?fld]]) :-  
    ComputedVarFieldCombo(?base1,?fld),  VarMustAlias(?base1,?base2,_).
    .plan 1:(2,1)

The join is fully efficient, but insertion of a few hundreds of thousands of tuples is not. I'm not sure if it's merely the tuple creation overhead or the need to insert them sorted.

-The above is a unary predicate on a record. In general, having a predicate indexed by a record value (constructed object) seems problematic. E.g., if we have a recursive record-type and try to compute a length predicate on it (which is pretty essential knowledge), it is simply impossible for any number of values over a few thousand.

Generally, I think that recursive record types are the main reason to have record types at all. I don't see how a solution based on "flattening" might work for recursive records. 

If this is not consistent with your understanding of how things should work, please let me know. Perhaps there is a workaround I'm overlooking.

**Comment by HerbertJordan** [comment]
You are right, the suggested flattening would only eliminate the performance issues of non-recursive records. Thanks for clarifying that recursive records are of major interest.

Your observations are fully consistent with what would be expected from the current implementation of records. There have once been plans for improved indexing of relations containing records, but those have never been realized due to human resource constraints.

I started working on an idea how to speed up the computation of the length of records, but I have to leave right now -- I will post an update when I got to it.

**Comment by HerbertJordan** [comment]
I was attempting to build a long-running test case for computing the length of recursive lists:
```
.decl letter( x : number )
letter(0).
letter(x+1) :- letter(x), x<20.

.type list = [ x : number , r : list ]
.decl lists( x : list )
lists(nil).
lists([x,nil]) :- letter(x).
lists([y,[x,l]]) :- lists([x,l]),letter(y),y<x.

.decl len( l : list, n : number )
len(nil,0).
len( [x,l] , n+1 ) :- lists([x,l]), len(l,n).

.printsize len ()
```
This code computes the length of a bit over 2M tuples within 3.6 seconds on my laptop (according to the profiler), which seems reasonable fast to me.

Can you provide a minimal example that exhibits record related performance issue?

**Comment by yanniss** [comment]
I'm afraid I cannot right now. In fact, I was about to post that I think the performance issues we are seeing are caused by a bug, not by the inefficient indexing. I'll follow up when I have more info, but right now it seems there is some recursive record value creation that shouldn't happen. I've gone through the relevant relations a few times and can't see anything wrong, but I can't find a bug in the output of the Souffle optimizer either. So, I'll just have to dig...

**Comment by b-scholz** [comment]
Anton wrote a deeply recursive n-queens example. That might be a good starting point for a benchmark.

**Comment by HerbertJordan** [comment]
Let's see what turns out to be the bottle neck in the real world example.

**Comment by b-scholz** [comment]
Yannis - I wanted to mention that the debug report (--debug-output=report.html) might be the best way to debug your problem, i.e., the RAM code may give you enough inside what is going on.

**Comment by yanniss** [comment]
I still don't have a minimal example, so let me draw instances of the problem from my full code. Here is a (simplified for readability) profiling entry:

     91m    .000     91m    2.37   7333136     0C353.12 AccessPathMayPointTo
AccessPathMayPointTo(?hctx,?heap,[?base2,[?fld,-]],?ctx,?insn) :-
  OptAccessPathWithVarMayPointTo(?hctx,?heap,?base1,?fld,?ctx,?insn),
  VarMustAlias(?base2,?base1,?ctx). 
 .plan 1:(2,1)

Clearly something is wrong. 91 minutes for a simple join that produces 7M tuples seems way too much. Let me make just one change, to use an intermediate relation:

Intermediate(?hctx,?heap,?base2,?fld,?ctx,?insn) :-
  OptAccessPathWithVarMayPointTo(?hctx,?heap,?base1,?fld,?ctx,?insn),
  VarMustAlias(?base2,?base1,?ctx).
 .plan 1:(2,1)

AccessPathMayPointTo(?hctx,?heap,[?base2,[?fld,-]],?ctx,?insn) :-
  Intermediate(?hctx,?heap,?base2,?fld,?ctx,?insn).

The result is *identical* but execution time drops to a (still unreasonable but much less so) 13mins. This suggests that the total cost of record object creation is affected by the presence of variables that participate in the join but not in the output (i.e., ?base1). This is perhaps not too surprising.

What surprises me is that the most expensive rule is not the second one but the first one:

     10m    .000     10m    2.31   7442504     0 C371.1 Intermediate
Intermediate (?hctx,?heap,?base2,?fld,?ctx,?insn) :-
  OptAccessPathWithVarMayPointTo(?hctx,?heap,?base1,?fld,?ctx,?insn),
  VarMustAlias(?base2,?base1,?ctx).  
 .plan 1:(2,1)

So, is this join slow? I suspect not! If I remove the second rule and just make "Intermediate" be an output relation, so that it's forced to be evaluated, the above join is instantaneous (and overall execution takes under 2min). Of course, in the latter case "Intermediate" is now no longer in a recursive cycle so it's evaluated without deltas, but the resulting size is not too different.

Basically, I suspect that the cost of the rule (the 10m version) is not in the join itself (i.e., computing the resulting tuples) but in updating an index into the other relation (AccessPathMayPointTo) which has a column that is of record type. But I may be wrong. It could be just that this join is much slower when done over deltas for some reason. Still, I strongly suspect that record values factor in somehow. I have not seen such performance issues without record values involved.

BTW, I do use the HTML report but I stop reading at the last Datalog code entry. I'll have to train myself to understand the RAM code. Soon...

**Comment by HerbertJordan** [comment]
Only looking at the total number of tuples in the resulting relation may provide an incomplete, maybe even missleading picture here. In you first rule
```
AccessPathMayPointTo(?hctx,?heap,[?base2,[?fld,-]],?ctx,?insn) :-
OptAccessPathWithVarMayPointTo(?hctx,?heap,?base1,?fld,?ctx,?insn),
VarMustAlias(?base2,?base1,?ctx).
```
it might very well be that the same combination of ?base2 and ?fld is attempted to be inserted several thousand times. Each time two tuples need to be created, which costs 2 memory accesses in a hash structure, thus two potential cache misses. 

Since in the second rule of your improved version only the actually needed 7M tuples are created, the execution time goes down to <3 minutes (you did not mention the execution time of this second rule).

However, the fact that your first rule takes the majority of time in your improved variant leads me to the suspicion that your VarMustAlias relation has a high fan-out. Since it is a alias relation, based on my experience I'd suggest that there are variables that are alias with several thousand other variables, although the majority has hardly any aliases. Thus, I'd suggest to investigate this fan-out as a likely source of your performance issues.


Another reason for the slow performance of the first rule in your improved variant might be this: due to some internal setup, relations with 6+ attributes may be significantly slower than relations with less attributes. The reason is that for presumable memory requirement issues starting at 6 attributes an indirect index is used instead of direct indices. However, you can override this by declaring your relation with a corresponding qualifier:
```
.decl relation_name ( ... ) btree
```
@b-scholz pls correct me if this syntax is wrong

**Comment by HerbertJordan** [comment]
For getting the fan-out under control something like this could help:
```
Aux(?base2,?base1,?ctx,?fld) :-
    VarMustAlias(?base2,?base1,?ctx),
    OptAccessPathWithVarMayPointTo(_,_,?base1,?fld,?ctx,_).

Intermediate(?hctx,?heap,?base2,?fld,?ctx,?insn) :-
    Aux(?base2,?base1,?ctx,?fld),
    OptAccessPathWithVarMayPointTo(?hctx,?heap,?base1,?fld,?ctx,?insn).
```
However, without the actual data set it is hard to say since I can't simply test it.

**Comment by yanniss** [comment]
Thanks for the suggestions! The 6+ attribute relation was something I was only vaguely aware of. I did resolve my performance problems. It was indeed an issue with variables having too many aliases and deriving the same tuples repeatedly. I solved it by computing representatives for alias classes.

But there was a reason why I cut my example where I did. This is the part that's pertinent to the topic of record value indexing. The second rule costs zero in my example. The first rule is identical to the slowest one (modulo tuple vs. two inline values), has 6 attributes (should be worse), and still costs 10 min vs. 91. Yes, I no longer have a problem in this case, but that 10 vs. 91 is the result of the lack of indexing for record values.

Admittedly, though, it's probably not a priority, and I have no real-world scenario where it hurts a lot.

**Comment by b-scholz** [comment]
If you have a close look at the profile log, you will find counters. These counters give you the amount of work that is unproductive in a rule. These counters start with # . In logic talk, they give you the number of failed proof-trees. These counters might give you another indication of how you can improve your spec. We used it way back for our DOOP port.

**Comment by Kostast23** [comment]
Could you please explain what the following counter tells us?
```#p-proof-counter;@new_Superinterface;/home/triantaf/doop-nexgen/out/1-object-sensitive+heap/xalan/1-object-sensitive+heap.dl [1492:1-1494:32];@new_Superinterface(?k,?c) :-     _DirectSuperclass(?c,?super),    @delta_Superinterface(?k,?super),    !Superinterface(?k,?c).;;5800```

**Comment by souffle-lang** [comment]
Every rule is converted to a loop-nest. It works like the following,

```
ctr=0
for t1 in R1: 
 if cond(t1) then: 
  for t2 in R2:
   if cond(t1,t2) then: 
      ....
   else: 
     ctr++ 
 else:
  ctr++
```

The counter is incremented every time when a loop/condition is not productive.

We should add this feature to the standard profiler and provide a standard measurement how productive a rule is in terms of searching. At the moment this is done for runtime only but not in terms of work units.
